pkg/controller/network/AzureFirewall/managed.go

```diff
@@ -0,0 +1,149 @@
+*/
+package AzureFirewall
+
+import (
+	"context"
+	azurenetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-06-01/network"
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/provider-azure/apis/network/v1alpha3"
+	azureclients "github.com/crossplane/provider-azure/pkg/clients"
+	azurefirewall "github.com/crossplane/provider-azure/pkg/clients/network"
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Error strings.
+const (
+	errNotAzureFirewall    = "managed resource is not an AzureFirewall"
+	errCreateAzureFirewall = "cannot create AzureFirewall"
+	errUpdateAzureFirewall = "cannot update AzureFirewall"
+	errGetAzureFirewall    = "cannot get AzureFirewall"
+	errDeleteAzureFirewall = "cannot delete AzureFirewall"
+)
+
+// Setup adds a controller that reconciles Security Group.
+func Setup(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha3.AzureFirewallKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha3.AzureFirewall{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha3.AzureFirewallGroupVersionKind),
+			managed.WithConnectionPublishers(),
+			managed.WithExternalConnecter(&connecter{client: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	client client.Client
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	creds, auth, err := azureclients.GetAuthInfo(ctx, c.client, mg)
+	if err != nil {
+		return nil, err
+	}
+	cl := azurenetwork.NewAzureFirewallsClient(creds[azureclients.CredentialsKeySubscriptionID])
+	cl.Authorizer = auth
+	return &external{client: cl}, nil
+}
+
+type external struct {
+	client azurenetwork.AzureFirewallsClient
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	v, ok := mg.(*v1alpha3.AzureFirewall)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotAzureFirewall)
+	}
+	az, err := e.client.Get(ctx, v.Spec.ResourceGroupName, v.Name)
+
+	if azureclients.IsNotFound(err) {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGetAzureFirewall)
+	}
+
+	if az.Name != nil {
+		azurefirewall.UpdateAzureFirewallStatusFromAzure(v, az)
+	}
```

Comment:
Do any spec fields need to be late initialized also?

Reply:
Can you please explain the statement we did not understand the context.

Reply:
Late initialization is the process of setting `spec` fields that are not provided by the user when they create the resource, but are set to default values by Azure. Here is an example of doing this: https://github.com/crossplane/provider-azure/blob/45f75a573ad4cc1727cfcee86d616b4892a11f74/pkg/clients/database/mysql.go#L306